client/relay-chain-network/src/lib.rs

```diff
@@ -0,0 +1,460 @@
+		Hash as PHash, Header as PHeader, InboundHrmpMessage,
+	},
+	InboundDownwardMessage, ParaId, PersistedValidationData,
+};
+use cumulus_relay_chain_interface::{RelayChainError, RelayChainInterface, RelayChainResult};
+use futures::{FutureExt, Stream, StreamExt};
+use jsonrpsee::{
+	core::client::{Client as JsonRPCClient, ClientT, Subscription, SubscriptionClientT},
+	rpc_params,
+	types::ParamsSer,
+	ws_client::WsClientBuilder,
+};
+use parity_scale_codec::{Decode, Encode};
+use polkadot_service::Handle;
+use sc_client_api::{StorageData, StorageProof};
+use sc_rpc_api::{state::ReadProof, system::Health};
+use sp_core::sp_std::collections::btree_map::BTreeMap;
+use sp_runtime::DeserializeOwned;
+use sp_state_machine::StorageValue;
+use sp_storage::StorageKey;
+use std::sync::Arc;
+
+pub use url::Url;
+
+const LOG_TARGET: &str = "relay-chain-network";
+const TIMEOUT_IN_SECONDS: u64 = 6;
+
+/// Client that maps RPC methods and deserializes results
+#[derive(Clone)]
+struct RelayChainRPCClient {
+	/// Websocket client to make calls
+	ws_client: Arc<JsonRPCClient>,
+
+	/// Retry strategy that should be used for requests and subscriptions
+	retry_strategy: ExponentialBackoff,
+}
+
+impl RelayChainRPCClient {
+	pub async fn new(url: Url) -> Self {
+		tracing::info!(target: LOG_TARGET, "Initializing RPC Client. Relay Chain Url: {}", url);
+		let ws_client = WsClientBuilder::default()
+			.build(url.as_str())
+			.await
+			.expect("Should be able to initialize websocket client.");
+
+		RelayChainRPCClient {
+			ws_client: Arc::new(ws_client),
+			retry_strategy: ExponentialBackoff::default(),
+		}
+	}
+
+	/// Call a call to `state_call` rpc method.
+	async fn call_remote_runtime_function(
+		&self,
+		method_name: &str,
+		hash: &PHash,
+		payload: Option<Vec<u8>>,
+	) -> RelayChainResult<sp_core::Bytes> {
+		let payload_bytes = payload.map_or(sp_core::Bytes(Vec::new()), sp_core::Bytes);
+		let params = rpc_params! {
+			method_name,
+			payload_bytes,
+			hash
+		};
+		self.request("state_call", params).await
+	}
+
+	/// Subscribe to a notification stream via RPC
+	async fn subscribe<'a, R>(
+		&self,
+		sub_name: &'a str,
+		unsub_name: &'a str,
+		params: Option<ParamsSer<'a>>,
+	) -> RelayChainResult<Subscription<R>>
+	where
+		R: DeserializeOwned,
+	{
+		self.ws_client
+			.subscribe::<R>(sub_name, params, unsub_name)
+			.await
+			.map_err(|err| RelayChainError::NetworkError(sub_name.to_string(), err))
+	}
+
+	/// Perform RPC request
+	async fn request<'a, R>(
+		&self,
+		method: &'a str,
+		params: Option<ParamsSer<'a>>,
+	) -> Result<R, RelayChainError>
+	where
+		R: DeserializeOwned + std::fmt::Debug,
+	{
+		retry_notify(
+			self.retry_strategy.clone(),
+			|| async {
+				self.ws_client
+					.request(method, params.clone())
+					.await
+					.map_err(|err| backoff::Error::Transient { err, retry_after: None })
+			},
+			|e, _| {
+				tracing::warn!("Unable to complete RPC request '{}', retrying. Error {}", method, e)
```

Comment:
```suggestion
				tracing::warn!(target: LOG_TARGET, error = %e, %method, "Unable to complete RPC request, retrying")
```

`tracing` provides these nice ways of "attaching" data to the log output.